lib/dao.js

```diff
@@ -629,6 +629,185 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          Model = Model.lookupModel(update);
+
+          var connector = self.getConnector();
+
+          // only when options.validate is not set, take model-setting into consideration
+          if (options.validate === undefined && Model.settings.automaticValidation === false) {
+            return callConnector();
+          }
+
+          inst.isValid(function(valid) {
+            if (!valid) return cb(new ValidationError(inst), inst);
+            callConnector();
+          }, update);
+
+          function callConnector() {
+            update = removeUndefined(update);
+            context = {
+              Model: Model,
+              where: where,
+              data: update,
+              currentInstance: inst,
+              hookState: ctx.hookState,
+              options: options
+            };
+            Model.notifyObserversOf('persist', context, function(err) {
+              if (err) return done(err);
+              connector.replaceOrCreate(Model.modelName, context.data, options, done);
+            });
+          }
+          function done(err, data, info) {
+            if (err) return cb(err);
+            var context = {
+              Model: Model,
+              data: data,
+              isNewInstance: info ? info.isNewInstance : undefined,
+              hookState: ctx.hookState,
+              options: options
+            };
+            Model.notifyObserversOf('loaded', context, function(err) {
+              if (err) return cb(err);
+
+              var obj;
+              if (data && !(data instanceof Model)) {
+                inst._initProperties(data, { persisted: true });
+                obj = inst;
+              } else {
+                obj = data;
+              }
+              if (err) {
+                cb(err, obj);
+              } else {
+                var context = {
+                  Model: Model,
+                  instance: obj,
+                  isNewInstance: info ? info.isNewInstance : undefined,
+                  hookState: hookState,
+                  options: options
+                };
+
+                Model.notifyObserversOf('after save', context, function(err) {
+                  if (!err) Model.emit('changed', inst);
+
+                  cb(err, obj, info);
+                });
+              }
+            });
+          }
+        }
+      });
+    } else {
+      var opts = {notify: false};
+      if (ctx.options && ctx.options.transaction) {
+        opts.transaction = ctx.options.transaction;
+      }
+      Model.findOne({where: ctx.query.where}, opts, function (err, inst){
+        if (err) return cb(err);
+        if (!isOriginalQuery) {
+          // The custom query returned from a hook may hide the fact that
+          // there is already a model with `id` value `data[idName(Model)]`
+          var pkName = idName(Model);
+          delete data[pkName];
+          if (inst) id = inst[pkName];
+        }
+        if (inst) {
```

Comment:
~I find this check a bit weird, isn't `inst` always set?~

Ah, `inst` is used in the callback arg for `findOne`, could you please rename it to something like `found` or `existingInst` to prevent confusion?

Now that I understand this better, I think my earlier comment about not allowing `if (inst) id = inst[pkName];` to change the `id` value is no longer relevant, I'll delete it.